File: Problem_2.js
// Time Complexity : O(m*n)
// Space Complexity : O(m*n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No


// Your code here along with comments explaining your approach
// First applied recursive solution and then applied bottom-to-top solution with 2D array for solving issue as taught in class.
/**
 * @param {number} amount
 * @param {number[]} coins
 * @return {number}
 */


var change = function (amount, coins) {
    //var memo = new Map();
    // //Recursion method with memoization
    // var changeHelper = (amount, coins, index, memo) => {
    //         if(amount < 0 || index == coins.length) return 0;
    //         if(amount == 0) return 1;
    //         if(memo.has(index+'-'+amount)) return memo.get(index+'-'+amount);

    //         let count1 = changeHelper(amount-coins[index], coins, index, memo)
    //         let count2 = changeHelper(amount, coins, index+1, memo)

    //         memo.set(index+'-'+amount, count1+count2);
    //         return memo.get(index+'-'+amount);
    // }

    // return changeHelper(amount, coins, 0, memo);
    let dp = new Array(coins.length + 1).fill(Array(amount + 1).fill(0));

    dp[0][0] = 1;
    for (let i = 1; i <= coins.length; i++) {
        dp[i][0] = 1;
    }
    for (let i = 1; i <= coins.length; i++) {
        for (let j = 1; j <= amount; j++) {
            if (coins[i - 1] > j) {
                dp[i][j] = dp[i - 1][j];
            }
            else {
                dp[i][j] = dp[i - 1][j] + dp[i][j - coins[i - 1]];
            }
        }
    }
    return dp[coins.length][amount];
};

File: problem2.java
// Time Complexity :O(m*n)
// Space Complexity :O(m*n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :no


// Your code here along with comments explaining your approach
// 1. first create a matrix for storing repeated sub problems ans.
// 2. keep the first row dummy.
// 3. use answers calculated previously to find the current prob;em answer.
// 4. keep in mind what are we asked and fill in valused accouring to the asked output.
// 5. also remeber what inputs we have, ddepending on that the intermediate dp data structure should be decided

class Solution {
    int result;
    public int change(int amount, int[] coins) {
        if(coins==null||coins.length==0) return 0;
        // return helper(coins,amount,0);
        int m = coins.length;
        int n = amount;
        int [][] dp = new int[m+1][n+1];
        dp[0][0]=1;
        for(int i =1; i<m+1; i++){
            for(int j = 0; j<n+1; j++){
                if(j<coins[i-1]) dp[i][j]=dp[i-1][j];
                else dp[i][j]=dp[i-1][j]+dp[i][j-coins[i-1]];
            }
        }
        return dp[m][n];

    }
    // private int helper(int[] coins, int amount, int index){
    //     //base
    //     if(amount == 0) return 1;
    //     if(amount<0 || index ==coins.length) return 0;
    //     //use
    //     int case1 = helper(coins, amount-coins[index], index);
    //     //notUse
    //     int case2 = helper(coins, amount, index+1);
    //     return case1+case2;
    // }
}